man/man1/snapccd.1
.TH SNAPCCD 1NEMO "30 January 1989"

.SH "NAME"
snapccd \- top view integrated velocity moment ccd-like image

.SH "SYNOPSIS"
.PP
\fBsnapccd in=\fPsnapshot \fBout=\fPimage [parameter=value]

.SH "DESCRIPTION"
\fBsnapccd\fP makes a ccd frame from an N-body snapshot, by
binning the (projected) x-y coordinates of all stars on a regular 
rectangular (square) grid and
assigning a mean integrated 'surface density' (integrated along z-axis, i.e. 
the line of sight,
The ccd frame is written to a standard \fIImage(5NEMO)\fP output file, and can
be accessed by various other programs for smoothing, display etc.
.PP
Stellar astronomers refer to this type of 2-d histogram
as a "Hess diagram" when it is used to plot a color-magnitude
diagram of stars. 
.PP
\fBsnapccd\fP can also create channel maps, where only signal near a
specified velocity is copied. See the \fBvrange\fP and \fBmoment\fP
keywords below. One can use either an exact velocity window, or
a gaussian beam in velocity space.
.PP
Note that the radial velocity has the astronomical convention. The observer
is located at infinity at the positive Z-axis: a positive z velocity
(\fBvz\fP) means stars are approaching, i.e. negative radial velocity.
.PP
\fIsnaprotate(1NEMO)\fP has to be used first to do any other than a top-view,
probably the more general version \fIsnapgrid(1NEMO)\fP should be used.
.PP
A better alternative is \fIsnapgrid(1NEMO)\fP, and for
more interpolation (as opposed to gridding) style is
\fIsnapmap(1NEMO)\fP


.SH "PARAMETERS"
.so/man1 parameters
.TP 20
\fBin=\fIin-file\fP
input file, must be in \fISnapShot(5NEMO)\fP format [no default].
.TP
\fBout=\fIout-file\fP
output file, will be in \fIImage(5NEMO)\fP format [no default].
.TP
\fBorigin=\fIx,y\fP
The origin \fI(x,y)\fP of the center of the ccd [default: \fB0,0\fP].
.TP
\fBsize=\fIvalue\fP
Full size of the CCD frame in 'physical' units. Note that the picture 
is forced to be square [default: \fB4.0\fP].
.TP
\fBcell=\fIvalue\fP
Cell- or pixel size. Again they will be square. For a 2D matrix
this value will be set by the header of that dataset.
[default: \fB0.1\fP].
.TP
\fBvrange=\fIvmin:vmax | vmean,vsig\fP
Range in (astronomical) radial velocity (\fB-vz\fP) to be used
only while binning the data. One can also give a mean and dispersion
in velocity space to weigh the data with. In this case the range in
velocity space is somewhat arbitrarely taken from
\fIvmean-3*vsig:vmean+3*vsig\fP. This last procedure is somewhat more
realistic for astronomical applications, unless the package where the
data will be transported to has decent smoothing utilities in velocity
space also. 
The default of this keyword is to take all data along the vz axis.
It does not make sense to select data along the z-axis when the
\fBmoment\fP (see below) is non-zero. [default: \fB-infinity:infinity\fP].
.TP
\fBmoment=\fInumber\fP
Order of the velocity moment: must be a non-negative number.
Most commonly choosen are: \fB0\fP (total intensity), \fB1\fP 
(velocity weighted intensity) and \fB2\fP (velocity square weighted intensity),
where 'intensity' should really be read as surface density per square unit length.
[default: \fB0\fP].

.SH "EXAMPLE"
The following example makes (three) CCD frames from an N-body snapshot,
smoothes the three moment maps
and combines them into an 'intensity' (\fBint\fP), 'mean velocity' (\fBvel\fP)
and 'velocity dispersion' (\fBsig\fP) map. 
.PP
Note that the moment maps must be smoothed \fIbefore\fP they
can be combined to the proper velocity and dispersion maps.
.nf
   % \fBsnapccd in=nbody.dat out=map0 moment=0\fP
   % \fBsnapccd in=nbody.dat out=map1 moment=1\fP
   % \fBsnapccd in=nbody.dat out=map2 moment=2\fP
   % \fBccdsmooth in=map0 out=map00 gauss=0.1\fP
   % \fBccdsmooth in=map1 out=map11 gauss=0.1\fP
   % \fBccdsmooth in=map2 out=map22 gauss=0.1\fP
   % \fBmv map00 int\fP
   % \fBccdmath in=int,map11     out=vel  fie=%2/%1\fP
   % \fBccdmath in=int,vel,map22 out=sig  fie="sqrt(%3/%1-%2*%2)"\fP
   % \fBrm map11 map22\fP
\fP
.fi

.SH "UNITS"
Units are maintained in the same way as in snapshots, they don't have
a specific name, but carry their normal meaning 'length', 'velocity'
and 'mass'. Since \fIsnapccd\fP calculates surface densities,
its units are formally 'mass' per square 'length' times 'velocity'
to the power \fBmoment\fP. 
.PP
When channel maps are produced (\fBmoment=0\fP), 
the data are not normalized w.r.t. the
convolving velocity beam. For a rectangular beam (\fBvrange=vmin:vmax\fP)
the data should formally be divided by \fB(vmax-vmin)\fP, for a 
gaussian beam (\fBvrange=vmean,vsig\fP) by \fBvsig*sqrt(2*pi)\fP.
Also remember that a gaussian beam has \fBFWHM = 2.355*sigma\fP.

.SH "FUTURE"
\fIsnapccd(1NEMO)\fP's function is rather similar to that of
\fIsnapplot(1NEMO)\fP,
except of plotting the position of the stars, its bins the data.
It would be a nice idea to combine both programs.

.SH "CAVEATS"
If smoothing is followed after gridding, the resulting beam is actually
slightly larger than the smoothing beam.

.SH "BUGS"
.PP
The program gracefully allows the user to use nonzero \fBmoment\fP
and a non-infinite \fBvrange\fP. No warnings here.

.SH "SEE ALSO"
snapgrid(1NEMO), snapmap(1NEMO),
snaprotate(1NEMO), snapslit(1NEMO), snapsmooth(1NEMO), snapaxsym(1NEMO), 
tsf(1NEMO), ccdgrid(1NEMO), ccdsky(1NEMO), ccdstack(1NEMO)
.PP
https://www.phosim.org/ or https://ascl.net/1307.011

.SH "AUTHOR"
Peter Teuben

.SH "FILES"
.nf
.ta +2.5i
src/nbody/image	snapccd.c, snapccd.1
.fi

.SH "HISTORY"
.nf
.ta +1.0i +4.0i
17-Jun-87	V1.0: Created	PJT
25-jun-87	V1.1: grayscale possibility added	PJT
29-jun-87	V2.0: image-format + three programs: ccd, smooth, look	PJT
30-Jun-87	V2.1: improved 'struct' filestructure	PJT
 1-Jul-87	V2.2: added velocity-moments option 	PJT
 8-jul-87	V2.4: proper defn. of cell position	PJT
 9-Mar-88	V3.0: added data history		PJT
 1-jun-88	V4.0: new filestruct, renamed programname	PJT
22-dec-88	V4.1: channel maps can be produces, keyword vrange	PJT
30-jan-89	V4.2: vel is now Zmin, also proper dimensions	PJT
.fi


src/nbody/image/snapccd.c
/* 
 *  SNAPCCD:  program makes a CCD frame from a snapshot
 *            see also SNAPGRID for a better alternative
 *
 *	16-Jun-87  V1.0 adapted from slit : 1D version   PJT
 *	17-Jun-87       dynamic allocation of 'frame'	 PJT
 *	25-jun-87  V2.0 write out image for further processsing   PJT
 *	30-jun-87  V2.1 new filestructure for image(5) PJT
 *	 6-jul-87  V2.3 roundoff controlled through floor()	PJT
 *	 8-jul-87  V2.4 correct implementation of cellposition   PJT
 *	20-aug-87  V3.0 better memory management in reading Nbody PJT
 *	 8-Mar-88  V3.1 added data history	PJT
 *	 1-jun-88  V4.0 rename name of file	PJT
 *			new filestruct; get/put_hist
 *	22-dec-88  V4.1 allow ranges in radial velocity space PJT
 *	18-jan-89  V4.2 compatible with 3D image structure PJT
 *	30-jan-89  V4.3 vel into Zmin	PJT
 *      16-mar-90  V4.4 made GCC happy, made helpvec  PJT
 *	13-nov-90  V4.5 new location of <snapshot.h>	PJT
 *       4-mar-97  V4.6 NEMO V2.x, fixes for SINGLEPREC pjt
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <snapshot/snapshot.h>
#include <image.h>
#include <history.h>

string defv[] = {
	"in=???\n			input filename (a snapshot)",
	"out=???\n			output filename (an image)",
	"origin=0,0\n			origin of CCD on 'sky'",
	"size=4.0\n			size (will be square)",
	"cell=0.0625\n			cellsize : 64 pixels if size=4",
	"vrange=-infinity:infinity\n	range in velocity space",
	"moment=0\n			velocity moment to weigh with",
	"VERSION=4.7a\n			9-oct-2023 PJT",
	NULL,
};

string usage = "simple mass-based conversion of snapshot to image";

#define EPS  0.00001
#ifndef HUGE
# define HUGE 1.0e20
#endif

local string	infile, outfile;			/* file names */
local stream  instr, outstr;				/* file streams */

		/* SNAPSHOT INTERFACE */
local int    nobj;
local real   tnow;
local real   *mass=NULL;
local real   *phase=NULL;
		/* IMAGE INTERFACE */
local imageptr  iptr=NULL;		/* will be allocated dynamically */
local int  nsize;		        /* map-size */
local real origin[2];			/* of map in sky coordinates */
local real size;			/* size of frame (square) */
local real cell;			/* cell or pixel size (square) */
local real xmin, ymin;                  /* lower left of grid */

local int    moment;	                /* moment to take in velocity */
local real vmin,vmax;			/* range in velocity space */
local real vmean, vsig;		        /* beam in velocity space */
local bool   vbeam;

local char *axname[] = {               /* axnames */
        "x", "y", "z", "vx", "vy", "vz"
    }; 


void setparams()
{
	string  tmpstr;
	int	tmpint;
	char    *cp;

	if (NDIM!=3)
		error("This program requires 3D-data, 2D could work");

	infile =  getparam ("in");
	outfile = getparam("out");

        tmpint = nemoinpr(getparam("origin"),origin,2);
	if (tmpint!=2)
	  error ("Keyword 'origin' must have two values, e.g. origin=0,0\n");

	size = getdparam("size");
	cell = getdparam("cell");
	nsize = rint(size/cell);    /* use nearest integer */
	dprintf (1,"size of CCD frame = %d square\n",nsize);

	moment=getiparam("moment");
	dprintf (1,"Vel_moment=%d\n",moment);
	if (moment<0)
		error("Vel-moments should be non-negative");

	tmpstr = getparam("vrange");
	cp = strchr(tmpstr,':');
	if (cp==NULL) {			/* beam specified:   vmean,vsig */
            cp = strchr(tmpstr,',');
            if (cp==NULL)
                error("syntax error vrange=vmin:vmax or vrange=vmean,vsig\n");
	    vmean = atof(tmpstr);
            vsig = atof(++cp);
            vmin = vmean - 3*vsig;
            vmax = vmean + 3*vsig;
            dprintf(2,"Gaussian beam specified: vmean=%f vsig=%f\n",vmean,vsig);
            vbeam = TRUE;
        } else {                    /* vrange specified: vmin:vmax */
            if (strncmp(tmpstr,"-infinity",9)==0)
                vmin = -HUGE;
            else
                vmin = atof(tmpstr);
            if (strncmp(++cp,"infinity",8)==0)
                vmax = HUGE;
            else
                vmax = atof(cp);
            dprintf(2,"vrange = %f : %f\n",vmin,vmax);
            vbeam = FALSE;
        }
}

int read_snap()
{				
    int    i;
    real   imass;    

    get_history(instr);

    if (! get_tag_ok(instr, SnapShotTag))
    	return(0);				/* no snapshot found */
    	
    get_set(instr, SnapShotTag);
      get_set(instr, ParametersTag);
        get_data(instr, NobjTag, IntType, &nobj, 0);
	if (get_tag_ok(instr, TimeTag))
	   get_data(instr, TimeTag, RealType, &tnow, 0);
	else
	   tnow=0.0;
      get_tes(instr, ParametersTag);
      dprintf (1,"nobj=%d time=%f\n", nobj, tnow);
      get_set(instr, ParticlesTag);
         if (mass==NULL) {		/* first time around */
            mass = (real *) malloc(nobj * sizeof(real));
            phase = (real *) malloc(nobj * 2 * NDIM * sizeof(real));
            if (mass==NULL || phase==NULL)
                error ("read_snap: not enough memory for N-body snapshot\n");
         }
         if (get_tag_ok(instr,MassTag))
           get_data(instr, MassTag, RealType, mass, nobj, 0);
         else {
	   dprintf (0,"WARNING: no masses provided, will assume 1/%d\n",nobj);
	   imass = 1.0/(real)nobj;
	   for (i=0; i<nobj; i++)
	      mass[i] = imass;
	 }
         get_data(instr, PhaseSpaceTag, RealType, phase, nobj, 2, NDIM, 0);
      get_tes(instr, ParticlesTag);
    get_tes(instr,SnapShotTag);
    return 1; 
}

void allocate_image()
{
    create_image (&iptr,nsize,nsize);       /* force square image */
    if (iptr==NULL) {
      dprintf (0,"Not enough memory for image\n");
      stop(0);
    }
    Nx(iptr) = Ny(iptr) = nsize;			/* image will be square */
    xmin = origin[0] - 0.5*size + 0.5*cell;   /* note definition of */
    ymin = origin[1] - 0.5*size + 0.5*cell;   /* cell position here */    
#if 0
    // old axis=0 mode
    Xmin(iptr) = xmin;
    Ymin(iptr) = ymin;
#else
    // new axis=1 mode
    Xmin(iptr) = origin[0];
    Ymin(iptr) = origin[1];
    Xref(iptr) = (nsize-1)/2.0;
    Yref(iptr) = (nsize-1)/2.0;
    Axis(iptr) = 1;
#endif
    Dx(iptr) = Dy(iptr) = cell;
    Dz(iptr) = 0.0;
    Frame(iptr) = (real *) malloc(Nx(iptr)*Ny(iptr)*sizeof(real));
    if (Frame(iptr)==NULL) {
      dprintf (0,"Not enough memory to allocate %d*%d frame array\n",Nx(iptr),Ny(iptr));
      stop(1);
    }
    Namex(iptr) = axname[0];        /* may change */
    Namey(iptr) = axname[1];
    Namez(iptr) = axname[5];
}

void bin_data()
{
    real xsky, ysky, vrad;
    real m_min, m_max, brightness, inv_surden, total;
    int  i, k, ix, iy, nx, ny, noutside, noutvel, ndata;
    real *pptr;
    
	/* (re)initialize CCD and some other local  variables */
    nx=Nx(iptr);
    ny=Ny(iptr);
    for (ix=0; ix<nx; ix++)
       for (iy=0; iy<ny; iy++)
	  MapValue(iptr,ix,iy) = 0.0;
    m_max = -HUGE;
    m_min = HUGE;
    inv_surden = 1.0 / (cell * cell);		/* scaling factor */
    noutside=noutvel=ndata=0;
    total=0.0;
		/* walk through all particles and accumulate ccd data */
    for (i=0, pptr=phase; i<nobj; i++) {
        xsky = *pptr++;			/* x */
        ysky = *pptr++;			/* y */
        pptr += NDIM;
        vrad = -(*pptr++);		/* v_z */
	ix = floor((xsky-xmin)/cell+0.5+EPS);	/* integer coords in CCD */
	iy = floor((ysky-ymin)/cell+0.5+EPS);
	if (ix<0 || iy<0 || ix>=nx || iy>=ny) {
	    noutside++;			/* not in CCD frame */
	    continue;
	}
        if (vrad<vmin || vrad>vmax) {
            noutvel++;
            continue;
        }
	brightness =   mass[i] * inv_surden;
        if (vbeam)
            brightness *= exp( -0.5*sqr((vrad-vmean)/vsig) );
	for (k=0; k<moment; k++)
	    brightness *= vrad;
	MapValue(iptr,ix,iy) +=   brightness;
     }  /*-- end particles loop --*/

	/* determine maximum in picture */
    for (ix=0; ix<nx; ix++)
       for (iy=0; iy<ny; iy++) {
       	  brightness = MapValue(iptr,ix,iy);
	  total += brightness;
	  if (brightness > m_max)
	  	m_max = brightness;
	  else if (brightness < m_min)
	  	m_min = brightness;
	  if (brightness!=0.0)
	  	ndata++;
	}

   
    MapMin(iptr) = m_min;               /* min and max of data */
    MapMax(iptr) = m_max;
    Beamx(iptr)=Beamy(iptr)= 0.0;       /* beam in position space */
    BeamType(iptr) = NONE;              /* no smoothing yet */
    if (vbeam) {
       Zmin(iptr) = vmean;               /* velocity of data */
       Beamz(iptr) = vsig;
    } else
       Zmin(iptr) = 0.5*(vmin+vmax);

    dprintf (1,"%d particles in CCD frame\n",nobj-noutside-noutvel);
    dprintf (1,"     (%d were outside frame, %d were not in velocity range)\n",
                    noutside, noutvel);
    dprintf (1,"%d cells contain non-zero data,  min and max in map are %f %f\n",
    		ndata, m_min, m_max);
    dprintf (1,"Total mass in map is %f\n",total*Dx(iptr)*Dy(iptr));

}

void nemo_main ()
{
	setparams();                    /* stuff command line [pars] */

	instr = stropen (infile, "r");
	outstr = stropen (outfile,"w");

	read_snap();			/* read N-body data */
	allocate_image();		/* make space for image */
	bin_data();			/* do the heavy work */
	write_image (outstr,iptr);	/* write the image */

	strclose(instr);
	strclose(outstr);
}
